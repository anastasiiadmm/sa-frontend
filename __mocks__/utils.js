import * as reduxHooks from 'react-redux';

jest.mock('../src/redux/store');
jest.mock('react-redux');

export const mockedUseSelectors = jest.spyOn(reduxHooks, 'useSelector');
export const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
