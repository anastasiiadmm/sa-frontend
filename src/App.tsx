import React, { lazy, Suspense, useCallback, useEffect } from 'react';
import { Route, Routes } from 'react-router-dom';

import Spinner from 'components/Spinner/Spinner';
import SignIn from 'containers/SignIn/SignIn';
import { useTokenConfigs } from 'hooks/useTokensConfigs';
import { IListener } from 'interfaces';
import { checkForTokens, clearTokens, logoutUser } from 'redux/auth/authSlice';
import { useAppDispatch } from 'redux/hooks';
import { deleteCookie, getCookie, nameRefreshCookies } from 'utils/addCookies/addCookies';
import {
  defaultLocalStorage,
  logoutLocalStorage,
  nameLocalStorage,
  userLocalStorage,
} from 'utils/addLocalStorage/addLocalStorage';

const AppRouter = lazy(() => import('AppRouter/AppRouter'));
const App: React.FC = () => {
  const dispatch = useAppDispatch();
  const tokenConfigs = useTokenConfigs();

  const initializeApp = useCallback(() => {
    const tokensLocal = userLocalStorage(getCookie(nameRefreshCookies));
    if (tokensLocal?.access && tokensLocal?.refresh) {
      dispatch(checkForTokens(tokensLocal));
    } else {
      deleteCookie(nameRefreshCookies);
      logoutLocalStorage();
      dispatch(clearTokens());
    }
  }, [dispatch]);

  const handleStorageEvent = useCallback(
    ({ key, newValue }: IListener) => {
      if (key === nameLocalStorage) {
        if (newValue === JSON.stringify(defaultLocalStorage) || newValue === null) {
          dispatch(logoutUser());
          logoutLocalStorage();
        } else {
          const tokensCopy = JSON.parse(newValue);
          tokensCopy.refresh = getCookie(nameRefreshCookies);
          dispatch(checkForTokens(tokensCopy));
        }
      }
    },
    [dispatch],
  );

  useEffect(() => {
    initializeApp();
  }, [initializeApp]);

  useEffect(() => {
    window.addEventListener('storage', handleStorageEvent);

    return () => {
      window.removeEventListener('storage', handleStorageEvent);
    };
  }, [handleStorageEvent]);

  return tokenConfigs ? (
    <Suspense fallback={<Spinner />}>
      <AppRouter />
    </Suspense>
  ) : (
    <Routes>
      <Route path='*' element={<SignIn />} />
    </Routes>
  );
};

export default App;
