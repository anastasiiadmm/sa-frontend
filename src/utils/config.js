export const apiURL = '';

export const defaultError = { detail: 'На сервере что-то пошло не так' };
