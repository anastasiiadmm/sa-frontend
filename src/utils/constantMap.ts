import L from 'leaflet';

import active from 'assets/images/icons/active_tracktor.svg';
import endB from 'assets/images/icons/endB.svg';
import inactive from 'assets/images/icons/inactive_tracktor.svg';
import startA from 'assets/images/icons/startA.svg';

export const duckIconStart = new L.Icon({
  iconUrl: startA,
  iconRetinaUrl: startA,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(30, 30),
  className: 'leaflet-div-icon',
});

export const duckIconEnd = new L.Icon({
  iconUrl: endB,
  iconRetinaUrl: endB,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(30, 30),
  className: 'leaflet-div-icon',
});
export const activeIcon = new L.Icon({
  iconUrl: active,
  iconRetinaUrl: active,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(28, 30),
  className: 'leaflet-div-icon',
});

export const inactiveIcon = new L.Icon({
  iconUrl: inactive,
  iconRetinaUrl: inactive,
  iconAnchor: new L.Point(14, 14),
  popupAnchor: new L.Point(14, 0),
  iconSize: new L.Point(28, 30),
  className: 'leaflet-div-icon',
});
