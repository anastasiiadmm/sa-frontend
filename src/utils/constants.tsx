export const pathsWithoutMargin = [
  '/open-map',
  '/field-climate',
  '/field-climate/station',
  '/field-climate/config',
];

export const climateOptions = [
  { value: 'last_communication', label: 'Последнее соединение' },
  { value: 'airTemp', label: 'Температура воздуха' },
  { value: 'soilTemp', label: 'Температура почвы' },
  { value: 'rh', label: 'Относительная влажность' },
  { value: 'rain24h', label: 'Осадки, 24ч' },
  { value: 'rain48h', label: 'Осадки, 2сут.' },
  { value: 'rain7d', label: 'Осадки, 7сут.' },
  { value: 'battery', label: 'Батарея' },
];

export const climateColors = {
  red: '%23ca3a3c',
  green: '%23107934',
  skyblue: '%231babe1',
  yellow: '%23f6c12d',
  orange: '%23ee741c',
  blue: '%230b75b9',
  gray: '%23d3d3d3',
};
