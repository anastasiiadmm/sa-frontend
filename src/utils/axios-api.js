import axios from "axios";
import {apiURL} from "utils/config";

export const axiosApi = axios.create({
  baseURL: apiURL,
});