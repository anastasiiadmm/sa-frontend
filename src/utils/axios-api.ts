import axios from 'axios';

import { checkForTokens, logoutUser } from 'redux/auth/authSlice';
import { apiURL } from 'utils/config';
import { logoutLocalStorage } from 'utils/token';
import store from 'redux/store';

const axiosApi = axios.create({
  baseURL: apiURL,
});

axiosApi.interceptors.request.use(async (config) => {
  const key = store.getState()?.auth?.tokens?.access;

  if (key) {
    config.headers = {
      ...config.headers,
      Authorization: `Bearer ${key}`,
      'Content-Type': 'application/json',
    } as any;
  }
  return config;
});

axiosApi.interceptors.response.use(
  async (config) => {
    return config;
  },
  async (error) => {
    const originalRequest = error.config;

    const statusCode = error?.response?.status;
    const index = store;
    const { tokens } = index.getState().auth;
    if (
      tokens &&
      statusCode === 401 &&
      error.config &&
      !error.config._isReady &&
      error.response.data.messages
    ) {
      originalRequest._isReady = true;
      try {
        const resp = await axiosApi.post('/accounts/refresh/', { refresh: tokens.refresh });
        if (resp.status === 200) {
          const newTokens = resp.data;
          axiosApi.defaults.headers.Authorization = `Bearer ${newTokens.access}`;
          const usersLocal = {
            user: index.getState()?.auth?.user,
            token: {
              access: resp.data.access,
              refresh: tokens.refresh,
            },
          };
          index.dispatch(checkForTokens(usersLocal));
          const obj: any = usersLocal;
          delete obj.token.refresh;
          localStorage.setItem('users', JSON.stringify(obj));
          window.dispatchEvent(new Event('storage'));
          return axiosApi(originalRequest);
        }
      } catch (e) {
        logoutLocalStorage();
        window.location.reload();
      }

      store.dispatch(logoutUser());
    }

    return Promise.reject(error);
  },
);

export default axiosApi;
