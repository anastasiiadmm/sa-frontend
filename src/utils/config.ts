import { AxiosRequestConfig, AxiosRequestHeaders } from 'axios';

import { checkForTokens, logoutUser } from 'redux/auth/authSlice';
import store from 'redux/store';
import { TSetupAxiosInterceptor } from 'type';

import { apiUrlsV2, serverUrls, serverUrlsSocket } from 'utils/constants';
import { logoutLocalStorage } from 'utils/token';

const { REACT_APP_ENVIRONMENT } = process.env;

const appEnvironment = REACT_APP_ENVIRONMENT || 'local';

export const apiURL2 = apiUrlsV2[appEnvironment];
export const socketApiSocket = serverUrlsSocket[appEnvironment];

export const apiUrlCrop = serverUrls[appEnvironment];

export const setupAxiosInterceptors: TSetupAxiosInterceptor = (axiosInstance) => {
  return axiosInstance.interceptors.request.use(async (config: AxiosRequestConfig) => {
    const key = store.getState()?.auth?.tokens?.access;

    if (key) {
      config.headers = {
        ...config.headers,
        Authorization: `Bearer ${key}`,
      } as AxiosRequestHeaders;
    }
    return config;
  });
};

export const setupAxiosInterceptorsResponse: TSetupAxiosInterceptor = (axiosInstance) => {
  return axiosInstance.interceptors.response.use(
    async (config) => {
      return config;
    },
    async (error) => {
      const originalRequest = error.config;

      const statusCode = error?.response?.status;
      const index = store;
      const { tokens } = index.getState().auth;
      if (
        tokens?.access &&
        statusCode === 401 &&
        error.config &&
        !error.config._isReady &&
        error.response.data.messages
      ) {
        originalRequest._isReady = true;
        try {
          const resp = await axiosInstance.post('/accounts/refresh/', { refresh: tokens.refresh });
          if (resp.status === 200) {
            const newTokens = resp.data;
            axiosInstance.defaults.headers.Authorization = `Bearer ${newTokens.access}`;
            const usersLocal = {
              access: resp.data.access,
              refresh: tokens.refresh,
              is_manager: tokens.is_manager,
            };
            index.dispatch(checkForTokens(usersLocal));
            const obj = {
              access: resp.data.access,
              is_manager: tokens.is_manager,
            };
            localStorage.setItem('users', JSON.stringify(obj));
            window.dispatchEvent(new Event('storage'));
            return axiosInstance(originalRequest);
          }
        } catch (e) {
          logoutLocalStorage();
          window.location.reload();
        }

        store.dispatch(logoutUser());
      }

      return Promise.reject(error);
    },
  );
};
