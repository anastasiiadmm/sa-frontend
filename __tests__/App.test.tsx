import {cleanup, render, screen} from '@testing-library/react';
import "@testing-library/jest-dom";
import '../__mocks__/matchMedia.mock'
import { mockedDispatch, mockedUseSelectors } from "../__mocks__/utils";

import {BrowserRouter} from 'react-router-dom';
import App from '../src/App';
import {act} from "react-test-renderer";

afterEach(cleanup);

jest.mock('react-leaflet', () => ({
    CircleMarker: () => null,
    MapContainer: () => null,
    Marker: () => null,
    Polyline: () => null,
    Popup: () => null,
    TileLayer: () => null,
}));

describe('App',  () => {

    test('App', async () => {
        mockedUseSelectors.mockReturnValue({
            tokenConfigs: true
        });
        const dispatch = jest.fn();
        mockedDispatch.mockReturnValue(dispatch);

        const { asFragment } = render(
            <BrowserRouter>
                <App />
            </BrowserRouter>
        );

        expect(asFragment).toMatchSnapshot();

        await act(async () => {});
    });


    test('renders App component', async () => {
        render(
            <BrowserRouter basename='/'>
                <App />
            </BrowserRouter>
        );
        expect(screen.getByTestId('sign-in')).toBeInTheDocument();
        await act(async () => {});
    });

})
