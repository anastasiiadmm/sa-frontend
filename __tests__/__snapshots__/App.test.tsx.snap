// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App 1`] = `[Function]`;
